src/instructlab/model/backends/backends.py

```diff
@@ -304,6 +305,75 @@ def shutdown_process(process: subprocess.Popen, timeout: int) -> None:
         )
         process.kill()
 
+    # Various facilities of InstructLab rely on multiple successive start/stop
+    # cycles. Since vLLM relies on stable VRAM for estimating capacity, residual
+    # reclamation activity can lead to crashes on start. To prevent this add a
+    # short delay (typically ~ 10 seconds, max 30) to verify stability.
+    #
+    # Ideally a future enhancement would be contributed to vLLM to more gracefully
+    # handle this condition.
+    wait_for_stable_vram(MAX_STABLE_VRAM_WAIT)
+
+
+def wait_for_stable_vram(timeout: int):
+    logger.info("Waiting for GPU VRAM reclamation...")
+    supported, stable = wait_for_stable_vram_cuda(timeout)
+    if not supported:
+        # TODO add support for intel
+        sleep(timeout)
+        return
+    if not stable:
+        # Only for debugging since recovery is likely after additional start delay
+        logger.debug(
+            "GPU VRAM did not stabilize during max timeout (%d seconds)", timeout
+        )
+
+
+def wait_for_stable_vram_cuda(timeout: int) -> Tuple[bool, bool]:
+    # Third Party
+    import torch
+
+    # Fallback to a constant sleep if we don't have support for the device
+    if not torch.cuda.is_available():
+        return False, False
+    start_time = time()
+    stable_samples = 0
+    last_free = 0
+    try:
+        while True:
+            free_memory = 0
+            try:
+                # TODO In the future this should be enhanced to better handle GPU partitioning.
+                # However, to do so will require that serve assign specific GPUs to vLLM, so
+                # that the same device pool is analyzed.
+                for j in range(torch.cuda.device_count()):
+                    device = torch.device(f"cuda:{j}")
+                    free_memory += torch.cuda.mem_get_info(device)[0]
+            except Exception:  # pylint: disable=broad-exception-caught
+                logger.warning(
+                    "Could not determine CUDA memory, falling back to general sleep"
+                )
+                return False, False
+
+            # Wait for recovery of free memory to stop for 5 successive samples ~ 10s.
+            # (6 passes counting init) Additional allocations are ignored since they
+            # may be spurious.
+            if last_free - free_memory >= 0:
+                stable_samples += 1
+                if stable_samples > 5:
```

Comment:
When everything is working properly (meaning the memory is freed up in < 6 cycles), how often does this return True, True vand return True, False because it didn't lower a 5th time?

Reply:
Not often, I only ever saw it occur artificially when i reduced the timeout value for the method. One case where it would occur is if someone has volatile activity on other GPUs outside of instructlab, then we could end up waiting for ever so thats where the max timeout comes in.

Reply:
After some quick testing, it seems to consistently make it to the 5th iteration and it's still freeing memory.  So my follow-up question is how is 5 enough to know it's stable?  Does the % not matter at all?